### PR TITLE
More `manual_map` improvements

### PR DESCRIPTION
**Pull request by Jarcho**
Hopefully there are no more false positives with this. Other lints that move an expression into a closure should also be checked at some point, there will likely be some of the same issues this lint had.

changelog: Suggest move closure in `manual_map` when needed
changelog: Fix false positive in `manual_map` where one local needs to be captured by reference, and another by value.

### CODE REVIEW

File: tests/ui/manual_map_option_2.rs [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+        None => None,
+    };
+
+    let s = String::new();
+    let x = 0u32;
+    // Can't use map, `s` is consumed, but `x` is borrowed.
+    let _: Option<(String, &u32)> = match Some(0) {
+        Some(_) => Some((f1(s), &x)),
+        None => None,
+    };
```
Comment (giraffate) [inline tests/ui/manual_map_option_2.rs:80]: The suggestion before the fix seems to work. Why does it emit no lint here?

File: tests/ui/manual_map_option_2.rs [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+    println!("{}", v[0]);
```
Comment (giraffate) [inline tests/ui/manual_map_option_2.rs:81]: Why it uses `v[0]` here?

File: tests/ui/manual_map_option_2.rs [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+
+    let s = String::new();
+    let s2 = String::new();
+    // Can't use map, `s` is consumed, but `s2` is used afterwards.
+    let _: Option<(String, &str)> = match Some(0) {
+        Some(_) => Some((f1(s), s2.as_str())),
+        None => None,
+    };
```
Comment (giraffate) [inline tests/ui/manual_map_option_2.rs:89]: The suggestion before the fix seems to work. Why does it emit no lint here?

File: tests/ui/manual_map_option_2.rs [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+    println!("{}", s2);
+
+    let s = String::new();
+    let s2 = String::new();
+    // Can't use map, `s` is consumed, but `s2` is used afterwards.
+    let _: Option<(String, &String)> = match Some(0) {
+        Some(_) => Some((f1(s), &s2)),
+        None => None,
+    };
```
Comment (giraffate) [inline tests/ui/manual_map_option_2.rs:98]: The suggestion before the fix seems to work. Why does it emit no lint here?

File: tests/ui/manual_map_option_2.rs [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+    println!("{}", s2);
+
+    let s = String::new();
+    let s2 = &mut String::new();
+    // Can't use map, `s` is consumed, but `s2` is used afterwards.
+    let _: Option<(String, &str)> = match Some(0) {
+        Some(_) => Some((f1(s), s2.as_str())),
+        None => None,
+    };
```
Comment (giraffate) [inline tests/ui/manual_map_option_2.rs:107]: The suggestion before the fix seems to work. Why does it emit no lint here?

File: tests/ui/manual_map_option_2.rs [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+    println!("{}", s2);
+
+    let s = String::new();
+    let v = vec![0];
+    // Can't use map, `s` is consumed, but `v` is used afterwards.
+    let _: Option<(String, u32)> = match Some(0) {
+        Some(i) => Some((f1(s), v[i])),
+        None => None,
+    };
```
Comment (giraffate) [inline tests/ui/manual_map_option_2.rs:116]: The suggestion before the fix seems to work. Why does it emit no lint here?

**Review by giraffate — COMMENTED** [review]
I made some comments on tests. Also, this updates `clippy_lints/src/entry.rs`, so can you add tests for `map_entry`?

**Comment by Jarcho** [comment]
I'm going to close this for now. Most of this is completely useless. Made a bad assumption about why something else broke.

**Pull request closed without merging**